x86/64: make i915 as a kmod with required firmware

joez opened the pull request:
i915 driver requires to load correct firmware to work on latest x86 GPU (not only iGPU but also dGPU), it is more reasonable to make it as a kernel module, so that initramfs is not required, and it can also save some space from the kernel image comparing being a built-in driver

This PR is mainly for #15380
See also: https://forum.openwrt.org/t/silly-little-i915-issue-on-x86/150405

joez committed:
x86: make i915 as a kmod with required firmware

i915 driver requires to load correct firmware to work on latest x86
GPU, it is more reasonable to make it as a kernel module, so that
initramfs is not required, and it can also save some space from the
kernel image comparing being a built-in driver

Signed-off-by: Joe Zheng <joe.zheng@intel.com>
Link: https://github.com/openwrt/openwrt/pull/16276
Signed-off-by: Robert Marko <robimarko@gmail.com>

1715173329 commented on target/linux/x86/64/config-6.6:
```diff
@@ -151,7 +151,7 @@ CONFIG_DRM_FBDEV_EMULATION=y
 CONFIG_DRM_GEM_SHMEM_HELPER=y
 # CONFIG_DRM_HYPERV is not set
-CONFIG_DRM_I915=y
+CONFIG_DRM_I915=m
```
```suggestion
```
Please drop this since it's handled by video.mk.

1715173329 replied:
I mean just completely remove `CONFIG_DRM_I915` from Kconfig.

joez replied:
I see, I can have a try. This is my first contribution to OpenWrt, are there any guidelines for this

joez replied:
removed `COMFIG_DRM_I915` from `config-6.6` and it works as expected, patch is updated

hauke commented on package/firmware/linux-firmware/intel.mk:
```diff
@@ -9,6 +9,16 @@ define Package/ibt-firmware/install
+define Package/i915-firmware/install
+	$(INSTALL_DIR) $(1)/lib/firmware/i915
+	$(CP) \
+		$(PKG_BUILD_DIR)/i915/*.bin \
```
The firmware files were already packaged in b6ac37110f19d130f608707b0bfcf8c1df60fbdf

joez replied:
I have just noticed this, it is great. This change allows for finer adjustments and is better suited for scenarios where storage space is limited.

In my use case, for example NAS, the storage is not an issue, and would like to support as many CPUs as possible, just like Ubuntu, so it is better to deploy all the available i915 as a whole.

hauke commented on target/linux/x86/64/config-6.6:
```diff
@@ -151,7 +151,6 @@ CONFIG_DRM_FBDEV_EMULATION=y
 CONFIG_DRM_FBDEV_OVERALLOC=100
 CONFIG_DRM_GEM_SHMEM_HELPER=y
 # CONFIG_DRM_HYPERV is not set
-CONFIG_DRM_I915=y
```
This should be constent over all x86 targets.
If you remove it here please remove it in the other places too:
```
target/linux/x86/64/config-6.6:CONFIG_DRM_I915=y
target/linux/x86/generic/config-6.6:CONFIG_DRM_I915=y
target/linux/x86/legacy/config-6.6:CONFIG_DRM_I915=y
```

joez replied:
I can do it, but I am not sure what are the generic and legacy subtargets used for, so just want to make sure that this change doesn't have any side effects.

hauke replied:
The generic and legacy target are used for 32 bit x86 CPUs.

I would like to keep the subtargets similar to reduce maintenance effort later on.

joez replied:
OK, let me update them too

joez replied:
done, check the latest PR

hauke commented on package/kernel/linux/modules/video.mk:
```diff
@@ -406,6 +406,25 @@ endef
+	+kmod-drm-display-helper +kmod-drm-buddy +kmod-acpi-video \
+	+LINUX_6_6:kmod-drm-exec +LINUX_6_6:kmod-drm-suballoc-helper
+  KCONFIG:=CONFIG_DRM_I915 \
+	CONFIG_FB_INTEL=n
```
When i915 is build for all as a module you can move `CONFIG_DRM_I915_CAPTURE_ERROR` and similar config options to this package too.

hauke commented on target/linux/x86/64/config-6.6:
```diff
@@ -151,7 +151,6 @@ CONFIG_DRM_FBDEV_EMULATION=y
 # CONFIG_DRM_HYPERV is not set
-CONFIG_DRM_I915=y
 CONFIG_DRM_I915_CAPTURE_ERROR=y
 CONFIG_DRM_I915_COMPRESS_ERROR=y
```
Please run `make kernel_oldconfig CONFIG_TARGET=subtarget` to update the configuration.

joez replied:
I have just tried this, but the result is a bit confusing, many unrelated kconfig options are changed
check the details in the attached [config-6.6.patch](https://github.com/user-attachments/files/16809250/config-6.6.patch)

joez replied:
I am trying to clean up first, and then apply i915 related ones

joez replied:
Because many of the kconfig iteams are defined in KernelPackage, it is not easy to make kernel_oldconfig, I will focus on the i915 related ones instead

hauke commented:
Is adding `intel-igpu-firmware-dmc` to an build using OpenWrt master not sufficient on recnet Intel boxes? Will the driver only look for the firmware file before the root fs is loaded?
 I think the GuC and HuC FW is not needed, you only need the `intel-igpu-firmware-dmc` FW files.

joez commented:
> Is adding `intel-igpu-firmware-dmc` to an build using OpenWrt master not sufficient on recnet Intel boxes? Will the driver only look for the firmware file before the root fs is loaded? I think the GuC and HuC FW is not needed, you only need the `intel-igpu-firmware-dmc` FW files.

For the latest iGPU the GuC and HuC are required for media and AI tasks, DMC is optional.
If the i915 driver is built-in, then the required firmware will failed to load because they are not in the rootfs at that time.

Here are the logs on N305 CPU for your reference:

```
[    3.190833] i915 0000:00:02.0: [drm] Transparent Hugepage support is recommended for optimal performance on this platform!
[    3.202749] i915 0000:00:02.0: Direct firmware load for i915/adlp_dmc.bin failed with error -2
[    3.211388] i915 0000:00:02.0: Falling back to sysfs fallback for: i915/adlp_dmc.bin
[   66.035122] i915 0000:00:02.0: [drm] *ERROR* GT0: GuC firmware i915/tgl_guc_70.bin: fetch failed -ETIMEDOUT
[   66.035123] i915 0000:00:02.0: [drm] Failed to load DMC firmware i915/adlp_dmc.bin. Disabling runtime power management.
[   66.035127] i915 0000:00:02.0: [drm] DMC firmware homepage: https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/i915
[   66.068700] i915 0000:00:02.0: [drm] GT0: GuC firmware(s) can be downloaded from https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/i915
[   66.084381] i915 0000:00:02.0: [drm] GT0: GuC firmware i915/tgl_guc_70.bin version 0.0.0
[   66.092562] i915 0000:00:02.0: [drm] *ERROR* GT0: GuC initialization failed -ENOEXEC
[   66.100311] i915 0000:00:02.0: [drm] *ERROR* GT0: Enabling uc failed (-5)
[   66.107107] i915 0000:00:02.0: [drm] *ERROR* GT0: Failed to initialize GPU, declaring it wedged!
[   66.116238] i915 0000:00:02.0: [drm:add_taint_for_CI] CI tainted:0x9 by intel_gt_set_wedged_on_init+0x2f/0x40
[   66.301885] [drm] Initialized i915 1.6.0 20201103 for 0000:00:02.0 on minor 0
```

And here are the logs on 8665 CPU:

```
[ 3.218930] i915 0000:00:02.0: Direct firmware load for i915/kbl_dmc_ver1_04.bin failed with error -2
[ 3.228133] i915 0000:00:02.0: Falling back to sysfs fallback for: i915/kbl_dmc_ver1_04.bin
[ 3.621520] i915 0000:00:02.0: [drm] [ENCODER:94:DDI A/PHY A] failed to retrieve link info, disabling eDP
[ 3.631296] i915 0000:00:02.0: [drm] [ENCODER:94:DDI B/PHY B] is disabled/in DSI mode with an ungated DDI clock, gate it
[ 3.660355] [drm] Initialized i915 1.6.0 20201103 for 0000:00:02.0 on minor 0
```

hauke commented:
As far as I understood Gen8 and older uses no firmware, gen9 has it optionally, but mostly disabled and gen12 needs DMC firmware for power saving and GuC for some 3D rendering. HuC firmware is only needed for additional media offloading.

The x86 target is normally not space constrained. I think many are operating their OpenWrt x86 devices headless, so the DMC firmware for power saving would be nice. I would add the GuC and HuC firmware just optimally to save some space. people can install it when they need 3D rendering or media encoding.

joez commented:
@hauke would you help to check the latest update?

hauke commented on package/firmware/linux-firmware/intel.mk:
```diff
@@ -216,6 +216,15 @@ define Package/e100-firmware/install
+define Package/i915-firmware/install
+	$(INSTALL_DIR) $(1)/lib/firmware/i915
+	$(CP) \
+		$(PKG_BUILD_DIR)/i915/*.bin \
```
This is packaging the same files as the `Package/intel-igpu-firmware*` packages. If the existing 3 are missing something please create a new package for it otherwise do not package the same file into to packages.

joez replied:
I understand you concern, the total size of all the firmwares is large for a router

```
linux-firmware-20240513/i915$ du -shc *_dmc* | tail -n1
996K    total
linux-firmware-20240513/i915$ du -shc *_guc* | tail -n1
17M     total
linux-firmware-20240513/i915$ du -shc *_huc* | tail -n1
8.8M    total
linux-firmware-20240513/i915$ du -shc * | tail -n1
28M     total
```

But take `amdgpu-firmware` for example, it is larger than `i915-firmware`, but still installed as a whole

```
linux-firmware-20240513$ du -sh amdgpu
83M     amdgpu
```

The reason I think, for the current use cases, if we want to enable GPU, whether for media, graphics or AI, the storage is not a problem for most cases.

It is a balance between simplicity and optimization. 

- Install all the available firmwares as a whole, can support all the platforms, not only the current one but also the future ones. The price to pay is the increased size.
- If we want to minimize the storage, we can customize to install the necessary ones only, even for the target platform only (e.g. ADL) and skip the others. The price to pay is the time to tune.

BTW. i915 is evolving rapidly, as well as the i915 firmware, it is not only for iGPU but also for dGPU now, more function will be added to support new platforms
e.g. `mtl_gsc_1.bin` is for [GSC Firmware](https://github.com/intel/igsc/blob/master/doc/introduction.rst) on MTL

```
linux-firmware-20240513/i915$ find . | grep -v dmc | grep -v guc | grep -v huc
./mtl_gsc_1.bin
```

1715173329 replied:
well the main problem is your new `i915-firmware` package will be conflict with `intel-igpu-firmware*`
probably you can create a new package called `intel-igpu-firmware-gsc` and depends on it, or remove `intel-igpu-firmware*` package.

joez replied:
Maybe we can consider to make `i915-firmware` as a meta package, which depends on the individual `intel-igpu-firmware-*`.
And can we change the name of `intel-igpu-firmware-*` to `i915-firmware-*`? The prefix `intel-igpu` is misleading, i915 driver/firmware is not for iGPU only, it is for dGPU too

1715173329 replied:
> Maybe we can consider to make `i915-firmware` as a meta package, which depends on the individual `intel-igpu-firmware-*`.

Well it would be a little bit hard to create a "real" meta package in OpenWrt. Adding all i915 firmware packages to dependencies is fine IMO.

> And can we change the name of `intel-igpu-firmware-*` to `i915-firmware-*`? The prefix `intel-igpu` is misleading, i915 driver/firmware is not for iGPU only, it is for dGPU too

Sounds good ;)

joez committed:
linux-firmware: rename packages for i915 firmware

Change the package name from intel-igpu-firmware-* to i915-firmware-*,
the prefix "intel-igpu" is misleading, i915 firmware is not only for
iGPU but also for dGPU now.

Remove the redundant "intel" as i915 is already well known.

More accurate file classification to handle following files correctly:

  adlp_dmc.bin
  mtl_huc.bin
  mtl_huc_gsc.bin
  mtl_gsc_1.bin

The pattern in regex is "([[:alnum:]]+)_([[:alnum:]]+)(_[\w-.]+)?\.bin",
where $1 is the platform, $2 is the firmware type (dmc, guc, huc, etc.),
and the optional $3 which is revision or other suffix.

Glob first to narrow down the target file set, and then split with "_"
to extract the firmware type (remove the ".bin" in case there is no $3)

Add package "i915-firmware" as a meta package to install all the i915
firmwares, it is a balance between simplicity and optimization.

* Installing all the available firmwares as a whole, can support all the
  platforms, not only the current one but also the future ones. The
  price to pay is the increased size.
* If we want to minimize the storage, we can customize to install the
  necessary ones only, even for the target platform only (e.g. ADL) and
  skip the others. The price to pay is the time to tune.

What I am going to do is:

* Let drm-i915 driver depend on i915-firmware-dmc, which is small and
  can cover most of the old platforms
* Let the user select i915-firmware to install all the i915 firmwares as
  a whole to cover the latest or future platforms

Signed-off-by: Joe Zheng <joe.zheng@intel.com>
Link: https://github.com/openwrt/openwrt/pull/16276
Signed-off-by: Robert Marko <robimarko@gmail.com>

joez commented:
@hauke @1715173329 The PR is updated according to our previous discussion, help to check

hauke commented on package/firmware/linux-firmware/intel.mk:
```diff
@@ -216,29 +216,49 @@ define Package/e100-firmware/install
-		$(1)/lib/firmware/i915/
+	for f in $(PKG_BUILD_DIR)/i915/*_huc*.bin; do                        \
+	  t=`echo $$$${f##*/} | cut -d_ -f2 | cut -d. -f1`;                  \
+	  if [ "$$$$t" = huc ]; then $(CP) $$$$f $(1)/lib/firmware/i915/; fi \
```
why do you need this for loop here?

joez replied:
The file name pattern in regex is `([[:alnum:]]+)_([[:alnum:]]+)(_\w+)?`, where `$1` is the platform name, `$2` is the firmware type (such as dmc, guc, huc, etc.), and the optional $3 which is revision or other suffix.
It is hard to handle such pattern with glob mechanism, so my solution is to glob it first to narrow down the target file set, and then split with `_` to extract the firmware type (remove the ".bin" in case there is no $3, e.g. "adlp_dmc.bin")

hauke replied:
Linux-firmware has these huc files:
```
i915/bxt_huc_2.0.0.bin
i915/bxt_huc_ver01_07_1398.bin
i915/bxt_huc_ver01_8_2893.bin
i915/cml_huc_4.0.0.bin
i915/dg1_huc_7.7.1.bin
i915/dg1_huc_7.9.3.bin
i915/dg1_huc.bin
i915/dg2_huc_gsc.bin
i915/ehl_huc_9.0.0.bin
i915/glk_huc_4.0.0.bin
i915/glk_huc_ver03_01_2893.bin
i915/icl_huc_9.0.0.bin
i915/icl_huc_ver8_4_3238.bin
i915/kbl_huc_4.0.0.bin
i915/kbl_huc_ver02_00_1810.bin
i915/mtl_huc_gsc.bin
i915/skl_huc_2.0.0.bin
i915/skl_huc_ver01_07_1398.bin
i915/tgl_huc_7.0.12.bin
i915/tgl_huc_7.0.3.bin
i915/tgl_huc_7.5.0.bin
i915/tgl_huc_7.9.3.bin
i915/tgl_huc.bin
```
Which of them do you want to package here?

joez replied:
All of these will be in the i915-firmware-huc, especially tgl_huc.bin and dg2_huc_gsc.bin

hauke replied:
why do you need this change?

joez replied:
> why do you need this change?

Can you be more specific?

If you are asking why need to classify the i915 firmwares into separated packages, check the [previous discussion](https://github.com/openwrt/openwrt/pull/16276#discussion_r1758520927)

If you are talking about the "for loop", here is my answer:

> The for loop is especally important to handle `mtl_huc_gsc.bin` correctly. Without it both `*_huc*.bin` and `*_gcs*.bin` will match it, which means both `i915-firmware-huc` and `i915-firmware-gsc` packages will contain `mtl_huc_gsc.bin`, it causes build error

More details in commit message of [linux-firmware: rename packages for i915 firmware](https://github.com/openwrt/openwrt/pull/16276/commits/63a93f04689934c627d9de40601bcea17911817b)

hauke commented on package/kernel/linux/modules/video.mk:
```diff
@@ -406,6 +406,48 @@ endef
+  DEPENDS:=@TARGET_x86 @DISPLAY_SUPPORT +kmod-backlight +kmod-drm-ttm \
+	+kmod-drm-ttm-helper +kmod-drm-kms-helper +kmod-i2c-algo-bit +i915-firmware-dmc \
+	+kmod-drm-display-helper +kmod-drm-buddy +kmod-acpi-video \
+	+LINUX_6_6:kmod-drm-exec +LINUX_6_6:kmod-drm-suballoc-helper
```
Support for older kernel version was remove, you can remove the `LINUX_6_6:` prefix..

joez replied:
updated, check the new patch

jtru commented:
So I've been trrying to follow up my own work in PR #16069 that proved insufficient for my own host to make the DMC firmware load,. and power saving to become properly active.

Do I understand correctly that, going forward, **i915** will be configured and built as a **module** for OpenWrt? Because if so, I will abandon my attempt at implementing firmware fallback loading support for the driver, which I've been musing about on IRC this weekend... :)


(I was originally going to file an issue with the following content, but I stumbled over this PR before submitting it. Still, I think it contains some useful analysis, and therefore I will post it - but you may skip it esp. if in the future, firmware loading for i915 on OpenWrt "just works"! :))


---

i915: firmware loading issues

A recently merged set of new packages for Intel i915 iGPU-related firmware files should have paved the way for proper power management enablement (i.e., lowest possible power consumption without actual utilization, to the tune of more than 1W saved on my ~7W consuming box) on OpenWrt for most Intel-based x86 hardware. Alas, there are a few bits missing to make these firmware files effective, once installed.

The root of the problem is a race condition between the built-in i915 kernel driver, and the rootfs-dependent availability of the firmware file(s). On my x86-based OpenWrt box, this manifests with kernel logs like this:

```
[...]
[    3.169076] i915 0000:00:02.0: Direct firmware load for i915/kbl_dmc_ver1_04.bin failed with error -2
[...]
[    4.226033] VFS: Mounted root (squashfs filesystem) readonly on device 8:2.
[...]
[   65.527667] i915 0000:00:02.0: [drm] Failed to load DMC firmware i915/kbl_dmc_ver1_04.bin. Disabling runtime power management.
[   65.539411] i915 0000:00:02.0: [drm] DMC firmware homepage: https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/i915
```

The *direct load* fails with ENOENT, because the kernel cannot access the filesystem that just isn't mounted yet when it starts looking after having initialized the i915 driver. I *guess* this problem would just go away if the i915 driver were configured as a module, and be loaded from the very same fs as the firmware files reside on, but I have not verified. I was wondering if that were an acceptable change to how x86-based OpenWrt images are configured by default before I tried to explore this any further.

As it turns out, as a potential alternative, there is a [fallback mechanism](https://docs.kernel.org/driver-api/firmware/fallback-mechanisms.html#firmware-kobject-uevent-fallback-mechanism) available in the kernel, and that could be made to work with OpenWrt (and which I did actually verify as working before posting this, albeit with an ugly hack in /etc/rc.local instead of doing it properly via hotplug).

The mechanism works like this on my machine with its given PCIe bus layout - when i915 loads and fails to find firmware, it established two new sysfs entries as hinted at in the above document:

```
/sys/devices/pci0000:00/0000:00:02.0/firmware/i915!kbl_dmc_ver1_04.bin/loading
/sys/devices/pci0000:00/0000:00:02.0/firmware/i915!kbl_dmc_ver1_04.bin/data
```

These encode which drvice (driver) requests a particular firmware file, and could, with some generalization (replacing instances of `!` with `/`, afaict), be made into a generic firmware-loading hook mechaism in OpenWrt's hotplug framework. There is some prior art involving the mechanism in [https://github.com/openwrt/openwrt/blob/b62e6f5beb2a574c67ba5159f0049e54cdf25d69/target/linux/ipq806x/base-files/etc/hotplug.d/firmware/11-ath10k-caldata#L9](11-ath10k-caldata) for ipq806x at least already, and I wonder if a similar approach should be taken for i915, at least if the related firmware packages are installed. All that's needed in the hotplug script is to signal that fw is about to be loaded by echoing into the "loading" file, then dumping the firmware data into the "data" file, and finally signalling that dumping data has completed by again writing into the "loading" file.

joez commented:
@jtru thanks for your information, especially the fallback firmware loading mechanism, we are facing the same issue, and there are always more than one way to do it. I prefer to build i915 driver as a kmod because:

* It is straight forward, the mechanism is well known because many Linux distributions such as Ubuntu use this way
* OpenWrt already supports kmod package very well, many packages such as `amdgpu` have already proved this
* Built as a kmod package can save space of the kernel image, now you can store the kmod packges to the external storage or install them on demand later

BTW. I have just noticed you are the author of PR #16069 👍. When I first worked on #16276 your patch had not been merged, so I just package all the required firmware for Intel GPU as a whole just like `amdgpu` does. Then I realize it is useful to separate them into smaller packages as you did, details in [previous discussion](https://github.com/openwrt/openwrt/pull/16276#discussion_r1749431115), help to review [linux-firmware: rename packages for i915 firmware](https://github.com/openwrt/openwrt/pull/16276/commits/63a93f04689934c627d9de40601bcea17911817b), I did some enhancement based on your solution.

jtru commented:
As long as the end result is that i915 can load any firmware it needs to enable powersaving on x86 OpenWrt, I'm all good ;) But it'd be great to know if the module-based approach is what the project is going forward with (I don't think there's any real downside to this method, ftr), or if the fallback mechanism with some userspace logic/support is going to be required because the driver will stay built-in. If the latter is the case, I would try to beat the workaround solution/hack I have right now for my own router into proper shape for upstream inclusion.

joez commented:
@hauke would you comment on this?

joez commented:
@hauke @robimarko help review and merge this if no any concern

jtru commented:
This should really make it into 24.10. Otherwise, a lot of x86-based OpenWrt hosts are going to needlessly waste a Watt or two for no good reason whatsoever.

robimarko commented:
Thanks! Rebased on top of main and merged!

robimarko commented:
Lets give it couple of days in master and if we dont get any regressions please ping me to backport to 24.10

namiltd commented:
Package kmod-drm-i915 is missing dependencies for the following libraries: intel-gtt.ko

pkgadd commented:
Hi

I'm seeing a side effect with this patch, which causes a regression (no display output) for me on 24.10.0-rc7 and current main on my ivy-bridge based Celeron 1037u board.

Previously i915 was always enabled statically in the kernel configs for x86_64, but with these changes to package it in a modular fashion, the default is not to include i915.ko on x86_64 anymore. As a result, my monitor (DVI-D connected to the HDMI port) remains blank after grub loads the kernel, breaking visual output. 

I think this needs a `default y if (x86_64 || i386)` for kmod-drm-i915.

```
# lspci -knn
00:00.0 Host bridge [0600]: Intel Corporation 3rd Gen Core processor DRAM Controller [8086:0154] (rev 09)
        Subsystem: Intel Corporation 3rd Gen Core processor DRAM Controller [8086:0154]
        Kernel driver in use: ivb_uncore
lspci: Unable to load libkmod resources: error -2
00:02.0 VGA compatible controller [0300]: Intel Corporation 3rd Gen Core processor Graphics Controller [8086:0156] (rev 09)
        DeviceName:  Onboard IGD
        Subsystem: Intel Corporation Device [8086:2010]
        Kernel driver in use: i915
00:14.0 USB controller [0c03]: Intel Corporation 7 Series/C210 Series Chipset Family USB xHCI Host Controller [8086:1e31] (rev 04)
        Subsystem: Intel Corporation 7 Series/C210 Series Chipset Family USB xHCI Host Controller [8086:1e31]
        Kernel driver in use: xhci_hcd
00:16.0 Communication controller [0780]: Intel Corporation 7 Series/C216 Chipset Family MEI Controller #1 [8086:1e3a] (rev 04)
        Subsystem: Intel Corporation 7 Series/C216 Chipset Family MEI Controller #1 [8086:1e3a]
00:1a.0 USB controller [0c03]: Intel Corporation 7 Series/C216 Chipset Family USB Enhanced Host Controller #2 [8086:1e2d] (rev 04)
        Subsystem: Intel Corporation 7 Series/C216 Chipset Family USB Enhanced Host Controller #2 [8086:1e2d]
        Kernel driver in use: ehci-pci
00:1b.0 Audio device [0403]: Intel Corporation 7 Series/C216 Chipset Family High Definition Audio Controller [8086:1e20] (rev 04)
        Subsystem: Intel Corporation 7 Series/C216 Chipset Family High Definition Audio Controller [8086:1e20]
00:1c.0 PCI bridge [0604]: Intel Corporation 7 Series/C216 Chipset Family PCI Express Root Port 1 [8086:1e10] (rev c4)
        Subsystem: Intel Corporation 7 Series/C216 Chipset Family PCI Express Root Port 1 [8086:1e10]
        Kernel driver in use: pcieport
00:1c.1 PCI bridge [0604]: Intel Corporation 7 Series/C210 Series Chipset Family PCI Express Root Port 2 [8086:1e12] (rev c4)
        Subsystem: Intel Corporation 7 Series/C210 Series Chipset Family PCI Express Root Port 2 [8086:1e12]
        Kernel driver in use: pcieport
00:1c.2 PCI bridge [0604]: Intel Corporation 7 Series/C210 Series Chipset Family PCI Express Root Port 3 [8086:1e14] (rev c4)
        Subsystem: Intel Corporation 7 Series/C210 Series Chipset Family PCI Express Root Port 3 [8086:1e14]
        Kernel driver in use: pcieport
00:1d.0 USB controller [0c03]: Intel Corporation 7 Series/C216 Chipset Family USB Enhanced Host Controller #1 [8086:1e26] (rev 04)
        Subsystem: Intel Corporation 7 Series/C216 Chipset Family USB Enhanced Host Controller #1 [8086:1e26]
        Kernel driver in use: ehci-pci
00:1f.0 ISA bridge [0601]: Intel Corporation HM77 Express Chipset LPC Controller [8086:1e57] (rev 04)
        Subsystem: Intel Corporation HM77 Express Chipset LPC Controller [8086:1e57]
        Kernel driver in use: lpc_ich
00:1f.2 SATA controller [0106]: Intel Corporation 7 Series Chipset Family 6-port SATA Controller [AHCI mode] [8086:1e03] (rev 04)
        Subsystem: Intel Corporation 7 Series Chipset Family 6-port SATA Controller [AHCI mode] [8086:1e03]
        Kernel driver in use: ahci
00:1f.3 SMBus [0c05]: Intel Corporation 7 Series/C216 Chipset Family SMBus Controller [8086:1e22] (rev 04)
        Subsystem: Intel Corporation 7 Series/C216 Chipset Family SMBus Controller [8086:1e22]
01:00.0 Ethernet controller [0200]: Intel Corporation 82574L Gigabit Network Connection [8086:10d3]
        Subsystem: Micro-Star International Co., Ltd. [MSI] Device [1462:6c40]
        Kernel driver in use: e1000e
02:00.0 Ethernet controller [0200]: Intel Corporation 82574L Gigabit Network Connection [8086:10d3]
        Subsystem: Micro-Star International Co., Ltd. [MSI] Device [1462:6c40]
        Kernel driver in use: e1000e
03:00.0 Network controller [0280]: MEDIATEK Corp. MT7921 802.11ax PCI Express Wireless Network Adapter [14c3:7961]
        Subsystem: AzureWave Device [1a3b:4680]
        Kernel driver in use: mt7921e
```

I have confirmed that rebuilding (main/ r28724-807074309d) with `CONFIG_PACKAGE_kmod-drm-i915=y` enabled explicitly fixes the problem for me. Without this, both 24.10.0-rc7 and main (r28608-g693c1ea81a) have no HDMI output for me.

@hauke this is a regression for openwrt-24.10 and main relative to openwrt-23.05 and earlier.

joez commented:
> Previously i915 was always enabled statically in the kernel configs for x86_64, but with these changes to package it in a modular fashion, the default is not to include i915.ko on x86_64 anymore. As a result, my monitor (DVI-D connected to the HDMI port) remains blank after grub loads the kernel, breaking visual output.

Yes, for backword compatibility we should keep the i915 as enabled by default for x86, maybe it can be implemented by adding:

```
DEFAULT:=y if TARGET_x86
```

On the other hand, it is a opportunity to think about which default value is the better one
